lib/streamaggr/streamaggr.go

```diff
@@ -891,20 +928,28 @@ func (a *aggregator) Push(tss []prompbmarshal.TimeSeries, matchIdxs []byte) {
 		// do not intern key because number of unique keys could be too high
 		key := bytesutil.ToUnsafeString(buf[bufLen:])
 		for _, sample := range ts.Samples {
+			a.muFlushAfter.Lock()
+			a.flushAfter.Update(float64(currentTimestamp - sample.Timestamp))
+			a.muFlushAfter.Unlock()
 			if math.IsNaN(sample.Value) {
 				a.ignoredNanSamples.Inc()
 				// Skip NaN values
+				a.ignoredNanSamples.Inc()
 				continue
 			}
 			if ignoreOldSamples && sample.Timestamp < minTimestamp {
 				a.ignoredOldSamples.Inc()
 				// Skip old samples outside the current aggregation interval
+				a.ignoredOldSamples.Inc()
 				continue
 			}
 			if maxLag < now-sample.Timestamp {
 				maxLag = now - sample.Timestamp
 			}
-			samples = append(samples, pushSample{
+			if ignoreOldSamples {
+				flushIdx = int((sample.Timestamp)/a.tickInterval+1) % aggrStateSize
```

Comment:
Does this calculation need `+1` here? I don't see any similar logic in `getFlushIndices` which should be consistent with this calculation, so it is not clear why this one uses `+1` here.

Wdyt about moving this logic into a separate function:
```golang
func getStateIDx(interval time.Duration, t time.Time) int64 {
	intervalMs := int64(interval / time.Millisecond)
	ts := t.UnixMilli()
	return (ts / intervalMs) % aggrStateSize
}
```
In this case `getFlushIndicies` will become:
```golang
func (a *aggregator) getFlushIndices(dedupInterval, interval time.Duration, dedupTime, flushTime time.Time) (int64, int64) {
	flushIndex := getStateIDx(interval, flushTime)
	dedupIndex := flushIndex

	if dedupInterval > 0 {
		dedupIndex = getStateIDx(dedupInterval, dedupTime)
	}
	return dedupIndex, flushIndex
}
```
And it will also be possible to use `getStateIDx` here so the logic will be consistent everywhere.

Reply:
getFlushIndices operates timestamps after the flush time, so it requires to truncate timestamp by interval, when in Push it should get a flush index of a timestamp, which should happen in a future

Reply:
updated a logic